File: bin/aggregate_evidence.R
#!/usr/bin/env Rscript

header <- c(
    "seq_name", "length", "assembly_cov", "genomad_virus_score",
    "genomad_fdr", "genomad_n_hallmarks", "genomad_marker_enrichment",
    "genomad_taxonomy", "map_numreads", "map_covbases", "map_coverage_pct",
    "map_meandepth", "viral_by_genomad"
)

args <- commandArgs(trailingOnly = TRUE)
fasta <- args[[1]]
virus_path <- args[[2]]
coverage_path <- args[[3]]

fasta_lines <- readLines(fasta)
fasta_headers <- fasta_lines[startsWith(fasta_lines, ">")]
seq_name <- sub("^>(\\S+).*", "\\1", fasta_headers)

length_match <- regexpr("length_([0-9]+)", seq_name, perl = TRUE)
length_values <- rep(NA_character_, length(seq_name))
length_values[length_match > 0] <- regmatches(seq_name, length_match)[length_match > 0]
length_values[length_match > 0] <- sub("^length_", "", length_values[length_match > 0])

cov_match <- regexpr("cov_([0-9.]+)", seq_name, perl = TRUE)
cov_values <- rep(NA_character_, length(seq_name))
cov_values[cov_match > 0] <- regmatches(seq_name, cov_match)[cov_match > 0]
cov_values[cov_match > 0] <- sub("^cov_", "", cov_values[cov_match > 0])

master <- data.frame(
    seq_name = seq_name,
    length = length_values,
    assembly_cov = cov_values,
    stringsAsFactors = FALSE,
    check.names = FALSE
)

virus <- read.delim(
    virus_path, sep = "\t", header = TRUE, check.names = FALSE,
    stringsAsFactors = FALSE
)
virus <- virus[, c("seq_name", "virus_score", "fdr", "n_hallmarks", "marker_enrichment", "taxonomy"), drop = FALSE]
names(virus)[-1] <- paste0("genomad_", names(virus)[-1])
virus$viral_by_genomad <- 1

coverage <- read.delim(
    coverage_path, sep = "\t", header = TRUE, check.names = FALSE,
    stringsAsFactors = FALSE, comment.char = ""
)
coverage <- coverage[, c("#rname", "numreads", "covbases", "coverage", "meandepth"), drop = FALSE]
names(coverage) <- c("seq_name", "map_numreads", "map_covbases", "map_coverage_pct", "map_meandepth")

merged <- merge(master, virus, by = "seq_name", all.x = TRUE, sort = FALSE)
merged <- merge(merged, coverage, by = "seq_name", all.x = TRUE, sort = FALSE)
merged <- merged[match(seq_name, merged$seq_name), , drop = FALSE]
merged$viral_by_genomad[is.na(merged$viral_by_genomad)] <- 0
merged <- merged[, header, drop = FALSE]

write.table(
    merged, file = "", sep = "\t", row.names = FALSE, col.names = TRUE,
    quote = FALSE, na = "NA"
)

File: bin/aggregate_evidence.py
#!/usr/bin/env python3
import csv
import re
import sys


HEADER = [
    "seq_name", "length", "assembly_cov", "genomad_virus_score",
    "genomad_fdr", "genomad_n_hallmarks", "genomad_marker_enrichment",
    "genomad_taxonomy", "map_numreads", "map_covbases", "map_coverage_pct",
    "map_meandepth", "viral_by_genomad",
]
VIRUS_COLUMNS = ["virus_score", "fdr", "n_hallmarks", "marker_enrichment", "taxonomy"]
COVERAGE_COLUMNS = ["numreads", "covbases", "coverage", "meandepth"]


def rows(path):
    with open(path, newline="") as handle:
        yield from csv.DictReader(handle, delimiter="\t")


def contigs(path):
    result = []
    with open(path) as handle:
        for line in handle:
            if line.startswith(">"):
                seq_name = line[1:].split(None, 1)[0]
                length = re.search(r"length_(\d+)", seq_name)
                assembly_cov = re.search(r"cov_([0-9.]+)", seq_name)
                result.append((seq_name, length.group(1) if length else "NA",
                               assembly_cov.group(1) if assembly_cov else "NA"))
    return result


def load(path, key, columns):
    result = {}
    for row in rows(path):
        result[row[key]] = {column: row.get(column, "") for column in columns}
    return result


def main():
    fasta, virus_path, coverage_path = sys.argv[1:]
    virus = load(virus_path, "seq_name", VIRUS_COLUMNS)
    coverage = load(coverage_path, "#rname", COVERAGE_COLUMNS)
    writer = csv.writer(sys.stdout, delimiter="\t", lineterminator="\n")
    writer.writerow(HEADER)
    for seq_name, length, assembly_cov in contigs(fasta):
        virus_row = virus.get(seq_name, {})
        coverage_row = coverage.get(seq_name, {})
        writer.writerow([
            seq_name, length, assembly_cov,
            *(virus_row.get(column, "NA") or "NA" for column in VIRUS_COLUMNS),
            *(coverage_row.get(column, "NA") or "NA" for column in COVERAGE_COLUMNS),
            1 if seq_name in virus else 0,
        ])


if __name__ == "__main__":
    main()
